// @omnicast/api/dist/toast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.showToast = exports.Toast = void 0;
const crypto_1 = require("crypto");
const bus_1 = require("./bus");
/**
 * A Toast with a certain style, title, and message.
 *
 * @example
 * ```typescript
 * import { showToast, Toast } from "@raycast/api";
 * import { setTimeout } from "timers/promises";
 *
 * export default async () => {
 *   const toast = await showToast({ style: Toast.Style.Animated, title: "Uploading image" });
 *
 *   await setTimeout(1000);
 *
 *   toast.style = Toast.Style.Success;
 *   toast.title = "Uploaded image";
 * };
 * ```
 */
class Toast {
    options;
    callbacks = {};
    id;
    /**
     * Deprecated - Use `showToast` instead
     */
    constructor(props) {
        this.id = `toast_${(0, crypto_1.randomBytes)(16).toString('hex')}`;
        this.options = {
            title: props.title,
            style: props.style ?? Toast.Style.Success,
            message: props.message
        };
        if (props.primaryAction) {
            const { onAction } = props.primaryAction;
            this.options.primaryAction = props.primaryAction;
            this.callbacks.primary = (0, bus_1.createHandler)(() => onAction(this));
        }
        if (props.secondaryAction) {
            const { onAction } = props.secondaryAction;
            this.options.secondaryAction = props.secondaryAction;
            this.callbacks.secondary = (0, bus_1.createHandler)(() => onAction(this));
        }
    }
    /**
     * The style of a Toast.
     */
    get style() { return this.options.style; }
    set style(style) { this.options.style = style; }
    /**
     * The title of a Toast. Displayed on the top.
     */
    get title() { return this.options.title; }
    set title(title) { this.options.title = title; }
    /**
     * An additional message for the Toast. Useful to show more information, e.g. an identifier of a newly created asset.
     */
    get message() { return this.options.message; }
    set message(message) { this.options.message = message; }
    /**
     * The primary Action the user can take when hovering on the Toast.
     */
    get primaryAction() { return this.options.primaryAction; }
    set primaryAction(action) { this.options.primaryAction = action; }
    /**
     * The secondary Action the user can take when hovering on the Toast.
     */
    get secondaryAction() { return this.options.secondaryAction; }
    set secondaryAction(action) { this.options.secondaryAction = action; }
    /**
     * Shows the Toast.
     *
     * @returns A Promise that resolves when the toast is shown.
     */
    async show() {
        const payload = {
            title: this.options.title,
            message: this.options.message,
            style: this.options.style,
        };
        if (this.options.primaryAction && this.callbacks.primary) {
            const { title, shortcut } = this.options.primaryAction;
            payload.primaryAction = {
                title, shortcut, onAction: this.callbacks.primary
            };
        }
        if (this.options.secondaryAction && this.callbacks.secondary) {
            const { title, shortcut } = this.options.secondaryAction;
            payload.secondaryAction = {
                title, shortcut, onAction: this.callbacks.secondary
            };
        }
        await bus_1.bus.request("toast.show", payload, { timeout: 1000 });
    }
    /**
     * Hides the Toast.
     *
     * @returns A Promise that resolves when toast is hidden.
     */
    async hide() {
        await bus_1.bus.request("toast.hide", {
            id: this.id
        }, { timeout: 1000 });
    }
    update;
}
exports.Toast = Toast;
(function (Toast) {
    /**
     * Defines the visual style of the Toast.
     *
     * @remarks
     * Use {@link Toast.Style.Success} for confirmations and {@link Toast.Style.Failure} for displaying errors.
     * Use {@link Toast.Style.Animated} when your Toast should be shown until a process is completed.
     * You can hide it later by using {@link Toast.hide} or update the properties of an existing Toast.
     */
    let Style;
    (function (Style) {
        Style["Success"] = "success";
        Style["Failure"] = "failure";
        Style["Animated"] = "animated";
    })(Style = Toast.Style || (Toast.Style = {}));
})(Toast || (exports.Toast = Toast = {}));
const showToast = async (init, title = "", message) => {
    const toast = typeof init === 'string' ? new Toast({ style: init, message, title }) : new Toast(init);
    await toast.show();
    return toast;
};
exports.showToast = showToast;
